engine/src/main/java/io/zeebe/engine/processing/common/EventTriggerBehavior.java

```diff
@@ -0,0 +1,279 @@
+import io.zeebe.engine.processing.deployment.model.element.ExecutableStartEvent;
+import io.zeebe.engine.processing.streamprocessor.MigratedStreamProcessors;
+import io.zeebe.engine.processing.streamprocessor.sideeffect.SideEffectQueue;
+import io.zeebe.engine.processing.streamprocessor.writers.StateWriter;
+import io.zeebe.engine.processing.streamprocessor.writers.TypedCommandWriter;
+import io.zeebe.engine.processing.streamprocessor.writers.Writers;
+import io.zeebe.engine.state.KeyGenerator;
+import io.zeebe.engine.state.instance.EventTrigger;
+import io.zeebe.engine.state.instance.StoredRecord.Purpose;
+import io.zeebe.engine.state.mutable.MutableElementInstanceState;
+import io.zeebe.engine.state.mutable.MutableEventScopeInstanceState;
+import io.zeebe.engine.state.mutable.MutableVariableState;
+import io.zeebe.engine.state.mutable.MutableZeebeState;
+import io.zeebe.protocol.impl.record.value.processinstance.ProcessInstanceRecord;
+import io.zeebe.protocol.record.intent.ProcessInstanceIntent;
+import io.zeebe.protocol.record.value.BpmnElementType;
+import java.util.function.ToLongFunction;
+import java.util.stream.Collectors;
+
+public class EventTriggerBehavior {
+  private final ProcessInstanceRecord eventRecord = new ProcessInstanceRecord();
+
+  private final KeyGenerator keyGenerator;
+  private final CatchEventBehavior catchEventBehavior;
+  private final TypedCommandWriter commandWriter;
+  private final StateWriter stateWriter;
+  private final MutableElementInstanceState elementInstanceState;
+  private final MutableEventScopeInstanceState eventScopeInstanceState;
+  private final MutableVariableState variablesState;
+
+  public EventTriggerBehavior(
+      final KeyGenerator keyGenerator,
+      final CatchEventBehavior catchEventBehavior,
+      final Writers writers,
+      final MutableZeebeState zeebeState) {
+    this.keyGenerator = keyGenerator;
+    this.catchEventBehavior = catchEventBehavior;
+    commandWriter = writers.command();
+    stateWriter = writers.state();
+
+    elementInstanceState = zeebeState.getElementInstanceState();
+    eventScopeInstanceState = zeebeState.getEventScopeInstanceState();
+    variablesState = zeebeState.getVariableState();
+  }
+
+  private ProcessInstanceRecord getEventRecord(
+      final ProcessInstanceRecord value,
+      final EventTrigger event,
+      final BpmnElementType elementType) {
+    eventRecord.reset();
+    eventRecord.wrap(value);
+    eventRecord.setElementId(event.getElementId());
+    eventRecord.setBpmnElementType(elementType);
+
+    return eventRecord;
+  }
+
+  public void unsubscribeFromEvents(final BpmnElementContext context) {
+    final var sideEffectQueue = new SideEffectQueue();
+    catchEventBehavior.unsubscribeFromEvents(context, commandWriter, sideEffectQueue);
+
+    // side effect can immediately executed, since on restart we not reprocess anymore the commands
+    sideEffectQueue.flush();
```

Comment:
I think it is not necessary to spent to much time here. We will remove the side effect stuff after the complete migration, since it is no longer necessary to differentiate it.

Reply:
Possibly - we will still need side effects to schedule timers, which will be done as soon as we subscribe to a timer. But it's only for subscription, so it's fine here :+1: